cloud/services/resourceskus/cache.go

```diff
@@ -0,0 +1,242 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package resourceskus
+
+import (
+	"context"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2020-06-01/compute"
+	"github.com/pkg/errors"
+	azure "sigs.k8s.io/cluster-api-provider-azure/cloud"
+)
+
+// Cache loads resource SKUs at the beginning of reconcile to expose
+// features available on compute resources. It exposes convenience
+// functionality for trawling Azure SKU capabilities. It may be adapted
+// to periodically refresh data in the background.
+type Cache struct {
+	client Client
+
+	// data is the cached sku information from Azure.
+	// synchronization required if data is cached across reconcile calls, (i.e., refreshed in background as Runnable via mgr.Add(...))
+	data []compute.ResourceSku
+}
+
+// NewCacheFunc allows for mocking out the underlying client
+type NewCacheFunc func(context.Context, azure.Authorizer) (*Cache, error)
+
+// NewCache instantiates a cache and initializes its contents.
+func NewCache(ctx context.Context, auth azure.Authorizer) (*Cache, error) {
```

Comment:
I left it inside New because I don't think it makes much sense to use an uninitialized cache and it also makes mocking trivial. If we move refresh to a separate func called by users we depend on clients to correctly invoke it (compare with the usage of the staticFn). I can see why that's a bit odd here, I agree it's not a great fit. 

Could we move it inside the controller/reconciler setup and pass the cache directly into the scope params instead?

Reply:
would probably move the NewCacheFunc into e.g. the controller and initialize it in main.go to make that work

Reply:
The context being passed in on New definitely helps to simplify the outer interface. Also, it is nice to not have to prime the cache after initialization.

What if the cache has a default time to live before refresh, it would then check last time updated before actions on the cache, internally. If an action on the cache will act on never refreshed or expired data, the cache calls refresh. Each action on the cache would then include a `context.Context` as the first parameter of the func. Wdyt?

Reply:
The sneaky part is there's no additional refresh right now :)

After I removed it from the manager, it's really just a list produced once at the start of reconcile. The extra Cache type is useful to provide functionality on top of the existing Azure types (which is...limited, IMO). That's why I don't particularly think making refresh public makes sense -- if anything, I'd prefer to convert it to e.g. a runnable, keep refresh private, and return the error from refresh to `runnable.Start()` (or something like that).

If we do that we can move cache initialization inside the Start func which takes a done channel. Need to do a little context/done channel mapping.

Reply:
Is that going to be more than you want to get into on this PR?

Reply:
for now, I'll probably just shuffle things a bit to make this a little nicer in the current form w.r.t context. I'll play with it and push an update once I think it's sane. Seeing the result size be ~10MB for a single subscription made me lean back a bit from caching across multiple subscriptions (which is implied by caching across reconciles), but that made it all the more valuable to only fetch + parse one time.

if we find this works well in practice we can always extend it. mostly feels like an easy/cheap way to reduce latency + extra http calls, and make feature discovery easier at the same time.

Reply:
Dang... 10MB is quite the payload. I almost want to include this in image creation or mount it as a file for the controller.

Reply:
We might be able to take the whole list and ignore restrictions. We'd need the restriction info per sub so we'd still need to make the same call, it doesn't help much.

Alternatively the Azure API does have some pagination if I'm not mistaken, so we could try to do something more clever around that in the caching stage, since we're already looking per-request. It helps for the machine case because we can terminate early, but for cluster failure domains we'd still need to iterate a larger set.

As I was writing this I remembered we aren't using the location filter anymore -- I bet that will help cut the size down quite a bit.

Reply:
I commented on the location and finished reading your comment at roughly the same time. Heh.

Reply:
okay, so depending on region looks to be ~600-775 KB for one sub. That's a lot more reasonable to cache over iterations. I'll take another pass an make this a background process again I think.

Reply:
I might leave this per request right now. I'd like to turn it into a global cache of all known Azure SKUs, with subscription specific-restrictions. Then the max cache size is basically static for all subs at the original 10MB, since the rest will be a much smaller map[string]Restriction{} with details on which subs restrict some resources.

Given this is already xxl and gets some decent improvement for other features, I can iterate